# My Second Hamoye Quiz

## initialize

In [138]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score


## Quiz data


In [4]:
df = pd.read_csv('energydata_complete.csv')
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

In [140]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

## 17.  From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the Root Mean Squared error in three D.P?

In [148]:
# If we don't split the dataset into train and test subsets, we get a simple model.
x = df[['T2']] 
y = df['T6']  
model = LinearRegression()
model.fit(x, y)
y_pred = model.predict(x)
rmse = np.sqrt(mean_squared_error(y, y_pred))
round(rmse,3)



3.644

In [149]:
# Or we can split the dataset into training and testing sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=2)

# Fit a linear regression model on the training data
model = LinearRegression()
model.fit(x_train, y_train)

# Make predictions on the testing data
y_pred = model.predict(x_test)

# Calculate the RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
round(rmse,3)

3.646

The answer is 3.644

## 18. Remove the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Normalize the dataset using the MinMaxScaler (Hint: Use the MinMaxScaler fit_transform and transform methods on the train and test set respectively). Run a multiple linear regression using the training set. Answer the following questions: What is the Mean Absolute Error (in three decimal places) for the  training set?


In [186]:
# Drop “date”, “lights” columns and make a new dataframe ndf
ndf=df.drop(['date','lights'], axis = 1)
features = ndf.drop(columns=['Appliances'])
target = ndf['Appliances']

# split the datasets into train and test sets
x_train, x_test, y_train, y_test = train_test_split(features, target, test_size=0.3, random_state=42)


# normalize data with MinMaxScaler
scaler = MinMaxScaler()
x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.transform(x_test)

# train the model 
model = LinearRegression()
model.fit(x_train_scaled, y_train)

# Make predictions  for y using the trained model on the training set
y_train_pred = model.predict(x_train_scaled)



In [187]:
# Make predictions  for y using the trained model on the training set
y_train_pred = model.predict(x_train_scaled)
# Calculate the Mean Absolute Error (MAE) for the training set
mae_train = mean_absolute_error(y_train, y_train_pred)
round(mae_train, 3)

53.742

## 19. What is the Root Mean Squared Error (in three decimal places) for the training set?


In [188]:
rmse_train = np.sqrt(mean_squared_error(y_train, y_train_pred))
round(rmse_train, 3) 


95.216

## 20. What is the Mean Absolute Error (in three decimal places) for test set?


In [189]:
# Make predictions  for y using the trained model on the test set
y_test_pred = model.predict(x_test_scaled)

# Calculate the Mean Absolute Error (MAE) for the test set
mae_test = mean_absolute_error(y_test, y_test_pred)
round(mae_test, 3)

53.643

## 21. What is the Root Mean Squared Error (in three decimal places) for test set?


In [190]:
rmse_test = np.sqrt(mean_squared_error(y_test, y_test_pred))
round(rmse_test, 3) 


93.64

## 22. Did the Model above overfit to the training set?


In [191]:
# To answer the questionn we shall compare metrics on train and test data - 
# if the metrics on test data are significantly worse then on train data, there is overfitting.
print(round(mae_train, 3), round(mae_test, 3))

53.742 53.643


In [170]:
print(round(rmse_train, 3), round(rmse_test, 3))

95.216 93.64


Both metrics are smaller for test data - that means, they are better(smaller the difference between the model predictions and real values). So, we can conclude that there is no overfitting!

# 23. Train a ridge regression model with default parameters. Is there any change to the root mean squared error (RMSE) when evaluated on the test set?

In [192]:
ridge_reg = Ridge()
ridge_reg.fit(x_train_scaled, y_train)


Ridge()

In [193]:
y_test_pred_r = ridge_reg.predict(x_test_scaled)
rmse_test_r = np.sqrt(mean_squared_error(y_test, y_test_pred_r))
round(rmse_test_r, 3) 


93.709

In [194]:
round(rmse_test, 3) 

93.64

RMSE on the test set with Ridge regression is almost the same(greater by only 0.06).

## 24. Train a lasso regression model with default value and obtain the new feature weights with it. How many of the features have non-zero feature weights?


In [195]:
lasso_reg = Lasso()
lasso_reg.fit(x_train_scaled, y_train)

feature_weights = lasso_reg.coef_
feature_weights

array([  0.        ,  56.94451601,   0.        ,  -0.        ,
         0.        ,   0.        ,  -0.        ,   0.        ,
        -0.        ,   0.        ,   0.        ,  -0.        ,
        -0.        ,  -0.        ,   0.        , -26.10288797,
        -0.        ,  -0.        ,   0.        ,  -0.        ,
       -50.29397595,   5.14314792,   0.        ,  -0.        ,
        -0.        ,  -0.        ])

In [196]:
#Count the number of features with non-zero feature weights
non_zero_feature_count = sum(feature_weights != 0)
non_zero_feature_count

4

## 25.What is the new RMSE with the Lasso Regression on the test set?


In [197]:
y_test_pred_l = lasso_reg.predict(x_test_scaled)
rmse_test_l = np.sqrt(mean_squared_error(y_test, y_test_pred_l))
round(rmse_test_l, 3) 

99.424